### ISSUE TITLE
[JENKINS-5323] within 1 executor: job started while post-actions of previous job still running

### ISSUE THREAD

**Issue by jenkins-infra-bot** [timja-org/jenkins-gh-issues-poc-11-14#7761]
<p>One executor on one machine: Finishes job A, starts the post-build-actions of job A and while they are still running starts job B.</p>

<p>We have one Hudson master and several Hudson clients. They all only have 1 executor per machine so on one machine officially only one job runs at a time.</p>

<p>Example of the timing of two subsequent jobs with all Hudson nodes offline except one (and remember we only have 1 executor per machine!):<br/>
0:59am: putting two jobs into the queue: Job1 and Job2<br/>
1:00am: Job1 starts ...<br/>
2:00am: Job1 finished executing all build steps<br/>
2:00am: Job1 post build actions start ...<br/>
2:01am: Job2 starts ...<br/>
2:03am: Job1 post build actions finished.<br/>
...</p>

<p>As you can see from 2:01am till 2:03am Job2 and the post-build-actions of Job1 are running in parallel on the same executor! That's the problem I am talking about.</p>

<p>We have the following post-build actions enabled:</p>
<ul>
	<li>Archive the artifacts</li>
	<li>Publish JUnit test result report (This is the one which tried to access files which were no longer there then.)</li>
	<li>E-mail Notification</li>
</ul>

<p>As far as I understand locks+latches won't do what we want to happen because we want to have Job1 and Job2 run in parallel in different executors/machines if there are more executors/machines online.</p>

<p>The first thing a job does is to delete all the workspaces on that machine except his own one. We do that because the clients disks are too small to hold all jobs workspaces. (Each jobs workspace is about 10 GB and we have about 100 jobs, but the disk of the Hudson machine is just 100 GB.)<br/>
But this deletion is actually just what made us realize about the parallelism.</p>

<p>The bug I want to report is that Job1-post-build-actions run in parallel with Job2 on the same machine with just one executor.</p>

<p>you might want to see the discussion on the users list regarding the issue:<br/>
<a href="http://jenkins.361315.n4.nabble.com/Post-actions-still-run-while-next-job-already-started-td1013097.html" class="external-link" target="_blank" rel="nofollow noopener">http://jenkins.361315.n4.nabble.com/Post-actions-still-run-while-next-job-already-started-td1013097.html</a></p>

---
<details><summary><i>Originally reported by <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=flofh">flofh</a>, imported from: <a href="https://issues.jenkins.io/browse/JENKINS-5323" target="_blank">within 1 executor: job started while post-actions of previous job still running</a></i></summary>
<i><ul>
<li><b>status</b>: Resolved
<li><b>priority</b>: Critical
<li><b>resolution</b>: Incomplete
<li><b>resolved</b>: 2014-06-14T18:08:59+00:00
<li><b>votes</b>: 2
<li><b>watchers</b>: 2
<li><b>imported</b>: 2025-11-14
</ul></i>
<details><summary>Raw content of original issue</summary>

<pre>
<p>One executor on one machine: Finishes job A, starts the post-build-actions of job A and while they are still running starts job B.</p>

<p>We have one Hudson master and several Hudson clients. They all only have 1 executor per machine so on one machine officially only one job runs at a time.</p>

<p>Example of the timing of two subsequent jobs with all Hudson nodes offline except one (and remember we only have 1 executor per machine!):
0:59am: putting two jobs into the queue: Job1 and Job2
1:00am: Job1 starts ...
2:00am: Job1 finished executing all build steps
2:00am: Job1 post build actions start ...
2:01am: Job2 starts ...
2:03am: Job1 post build actions finished.
...</p>

<p>As you can see from 2:01am till 2:03am Job2 and the post-build-actions of Job1 are running in parallel on the same executor! That's the problem I am talking about.</p>

<p>We have the following post-build actions enabled:</p>
<ul>
	<li>Archive the artifacts</li>
	<li>Publish JUnit test result report (This is the one which tried to access files which were no longer there then.)</li>
	<li>E-mail Notification</li>
</ul>

<p>As far as I understand locks+latches won't do what we want to happen because we want to have Job1 and Job2 run in parallel in different executors/machines if there are more executors/machines online.</p>

<p>The first thing a job does is to delete all the workspaces on that machine except his own one. We do that because the clients disks are too small to hold all jobs workspaces. (Each jobs workspace is about 10 GB and we have about 100 jobs, but the disk of the Hudson machine is just 100 GB.)
But this deletion is actually just what made us realize about the parallelism.</p>

<p>The bug I want to report is that Job1-post-build-actions run in parallel with Job2 on the same machine with just one executor.</p>

<p>you might want to see the discussion on the users list regarding the issue:
<a href="http://jenkins.361315.n4.nabble.com/Post-actions-still-run-while-next-job-already-started-td1013097.html" class="external-link" target="_blank" rel="nofollow noopener">http://jenkins.361315.n4.nabble.com/Post-actions-still-run-while-next-job-already-started-td1013097.html</a></p></pre>
</details>
</details>
<details><summary><i>environment</i></summary>

```
Win2003 64-bit, with 64-bit JVM, Hudson 1.338<br/>
Win7 64-bit, Jenkins 1.514, Execute concurrent builds enabled
```
</details>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/danielbeck.png" title="danielbeck's avatar" /> danielbeck's <a href="https://issues.jenkins.io/browse/JENKINS-5323?focusedId=202730&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-202730">comment</a>:</i></summary>


<details><summary><sub><i>Raw content of original comment:</i></sub></summary>

<pre>
<p>Does this issue still occur on recent Jenkins versions? If so, please provide updated information.</p>

<p>Also, couldn't you just use the Workspace Cleanup plugin to clean up after every job execution?</p>
</pre>
</details>

</details>
<p>Does this issue still occur on recent Jenkins versions? If so, please provide updated information.</p>

<p>Also, couldn't you just use the Workspace Cleanup plugin to clean up after every job execution?</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/danielbeck.png" title="danielbeck's avatar" /> danielbeck's <a href="https://issues.jenkins.io/browse/JENKINS-5323?focusedId=203871&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-203871">comment</a>:</i></summary>


<details><summary><sub><i>Raw content of original comment:</i></sub></summary>

<pre>
<p>Resolving as 'Incomplete' after no response to comment asking for updated information in two weeks.</p>
</pre>
</details>

</details>
<p>Resolving as 'Incomplete' after no response to comment asking for updated information in two weeks.</p>
